src/repair.rs

```diff
@@ -54,26 +57,30 @@ pub enum RepairRequest {
 /// If well-formed, it thus contains the corresponding variant of [`RepairRequest`].
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub enum RepairResponse {
-    /// Response with the total number of slices for a specific block.
-    SliceCount(RepairRequest, usize),
+    /// Response with the last slice's Merkle root hash, plus corresponding proof.
+    LastSliceRoot(RepairRequest, SliceIndex, Hash, Vec<Hash>),
     /// Response with the Merkle root hash of a specific slice, plus corresponding proof.
     SliceRoot(RepairRequest, Hash, Vec<Hash>),
     /// Response with a specific shred.
     Shred(RepairRequest, Shred),
-    // TODO: remove or replace with variant that includes proof
-    Parent(RepairRequest, Slot, Hash),
 }
 
 /// Instance of double-Merkle based block repair protocol.
 pub struct Repair<N: Network> {
     blockstore: Arc<RwLock<Box<dyn Blockstore + Send + Sync>>>,
     pool: Arc<RwLock<Box<dyn Pool + Send + Sync>>>,
+
+    slice_roots: BTreeMap<(BlockId, SliceIndex), Hash>,
+    outstanding_requests: BTreeMap<Hash, RepairRequest>,
+    request_timeouts: BinaryHeap<(Instant, Hash)>,
+
     network: N,
     repair_channel: (
         tokio::sync::mpsc::Sender<BlockId>,
-        tokio::sync::Mutex<tokio::sync::mpsc::Receiver<BlockId>>,
+        Option<tokio::sync::mpsc::Receiver<BlockId>>,
```

Comment:
i find this a bit un-ergonomic.  But I think we can leave this is as is for this PR and explore ways of improving it subsequently.